include/dali/core/permute.h

```diff
@@ -85,6 +86,25 @@ OutPerm inverse_permutation(const InPerm &permutation) {
   return inv_perm;
 }
 
+/**
+ * @brief Permutes a sequence in place
+ *
+ * @remarks The function will return incorrect result or hang if `idx` contains repetitions.
+ */
+DALI_NO_EXEC_CHECK
+template <typename Seq, typename Indices>
+DALI_HOST_DEV
+void permute_in_place(Seq &inout, Indices &&idx) {
+  using index_type = std::remove_reference_t<decltype(idx[0])>;
+  for (index_type i = 0, n = dali::size(idx); i < n; i++) {
+      index_type src_idx = idx[i];
+      while (src_idx < i)
+          src_idx = idx[src_idx];
+      if (src_idx != i)
+          cuda_swap(inout[i], inout[src_idx]);
+  }
+}
```

Comment:
This is a part of another PR, I'll rebase when CI is operational and it's merged.